### DISCUSSION TITLE
"Disco effect" with SK6812 even with external power supply

### DISCUSSION THREAD

**Discussion by LeVraiRoiDHyrule** [awawa-dev/HyperHDR#557] [Q&A]
Hi,
I'm new to HyperHDR. I was planning to use Hyperion at first but as their Docker container are a complete mess with missing libraries and settings, I went to HyperHDR.
I am planning to control ~250 SK6812 5V leds. Video capture already works fine with an USB capture card.

At first I tried to control them directly from my Raspberry Pi 4, on the SPI bus. I had a "disco effect" with white flashes, and incorrect colors. So I used this https://wled-calculator.github.io/ and found out 7A are needed for 250 leds to work. So I bought a 10A 5V power supply. Today I've tried it. I powered the leds with it, the Raspberry still using its own power supply, and I made the GNDs common. I still had the "disco effect" and no correct colors. I know that SPI control directly from the pi is not officially supported, but I would like to try that way first to begin using HyperHDR, and then moving to a more complex solution.

If I understand correctly from https://www.hyperhdr.eu/2023/02/ultimate-guide-on-how-to-build-led.html#chapter2c, HyperSPI is recommended over direct SPI. I also read that a 3.3 to 5V level shifter is needed for Raspberries and D1 mini cards. I find that strange, because I also have a lolin D1 mini that also controls another set of SK6812 in another room (no Hyperion, just WLED), and I didn't need any level shifter. I've plugged the signal directly on the D1 mini and it is working. So I have multiple questions:

- Why, in the case of HyperHDR compared to simple WLED, a level shifter is needed?
- What could cause my "disco effect" problem ? Is it from SPI communication of from the absence of level shifter ? Or something else ?
- Is it possible to help me do basic SPI before doing another way like HyperSPI once this works correclty ?
- Am I starting on the right foot ? Any advice or thing I'm doing wrong ?

I'm a bit lost because this is a lot of new things for me, and would greatly appreciate some help. Anyway, HyperHDR looks awesome. Thanks in advance for any answer, have a great day !

**Comment by awawa-dev** [comment]
Hi
The LEDs can operate without a level shifter for some setups when the data line only uses 3.3V instead of 5V, but that's definitely outside of the LED's technical specification. It is not guaranteed not to flash when the voltage fluctuates and occasionally jumps beyond that narrow margin of tolerance causing the LED to flash or freeze.  

HyperSPI uses Raspberry Pi and ESP32/ESP8266: they all use **3.3V** logic level, so they can be used that way. SPI LEDs (not SK6812, it's neopixel) are supported when driven directly from the Raspberry Pi, but only when the appropriate level shifter is used.   

There are two most common causes of the described effect: missing 3.3V to 5V level shifter or a problem with a common ground First verify the common ground, make sure there is an electrical path between two grounds using a multimeter , the common ground should be **short** similar as the data/clock cables to the LED strip, simple routed so do not to create any loops with it. If that doesn't help, use the appropriate level shifter.

**Reply by awawa-dev** [reply]
And if you are trying to use SPI for sk6812 that usually ends bad: SPI LED's data line frequency doesn't need to be perfect for SPI leds ... usually it fluctuates a bit but so the secondary clock line does and it synchronizes both and finally it can be read without a problem. But for Neopixel LEDs (when SPI bitbang is used) such fluctuation is not acceptable because it doesn't have a clock line to correct it.

**Reply by LeVraiRoiDHyrule** [reply]
> And if you are trying to use SPI for sk6812 that usually ends bad: SPI LED's data line frequency doesn't need to be perfect for SPI leds ... usually it fluctuates a bit but so the secondary clock line does and it synchronizes both and finally it can be read without a problem. But for Neopixel LEDs (when SPI bitbang is used) such fluctuation is not acceptable because it doesn't have a clock line to correct it.

Thanks a lot for your very detailed and quick answer, it really means a lot.

It's the 1st time I'm hearing about Neopixel, so I guess my leds are basic SK6812 and not Neopixel.

One thing I forgot to add is that even though I have flashing and incorrect colors, the number of leds that light up is correct and reflects my settings. 

I checked and my grounds are common, so I guess the only left culprit is the signal not being 5V, right? But if I understood correctly, SK6812 have some sort of tolerance?
if the only thing left to try is the level shifter, could you help me to choose it? Do you have a reference I should buy?

I find strange that the same setup works correctly for my WLED on D1 mini. 

Thanks in advance for any answer, have a great day.

**Reply by LeVraiRoiDHyrule** [reply]
I just found out about HyperSerialWLED. Although, if I understand correctly, it is slower than HyperSPI as it uses USB, I think this might be a good option because I also have 1m of led that will be used as normal lighting (not for HyperHDR) and that would allow me to put everything on the same device (I will go for another D1 mini I think). Then if later I move HyperHDR to a device that doesn't have GPIO, it won't be a problem.

**Reply by awawa-dev** [reply]
You can use HyperSerialWLED or HyperSerialEsp8266 (or HyperSerialESP32 for esp32/esp32-s2 mini) but your esp8266 (d1 mini?) board must have real ch340g chip to achieve 2Mb speed. These days most of these boards are equipped with poorly counterfeited chips that are unable to use 2Mb (usually you can still use them but you must recompile the firmware to use slower 921600 instead of 2000000). Currently esp32 or esp32-s2 are recommended as a safer and better solution.

> I just found out about HyperSerialWLED. Although, if I understand correctly, it is slower than HyperSPI as it uses USB,

Your LED strip is very small, the difference is negligible.

> It's the 1st time I'm hearing about Neopixel, so I guess my leds are basic SK6812 and not Neopixel

Your sk6812 is a Neopixel LED strip type, the protocol is similar to ws2812b with some nuances (rgbw vs rgb, a bit different reset timing).
Neopixel protocol is poorly emulated using Rpi's SPI protocol bitbanging. ESP boards have specialized tasks to focus on the LED rendering, do not must care about the OS handling routines like Rpi, and they're using different hardware acceleration methods like UART or I2S that work much better for Neopixel. 

If you read the FAQ there is one more solution to emulate sk6812 on Rpi that behaves better than SPI but it requires root privileges. You can try it on your own, do not recommend or support either of them.

**Comment by LeVraiRoiDHyrule** [comment]
> must have real ch340g chip to achieve 2Mb speed

Very interesting, thanks a lot for all those infos. I found out that I have an old D1 mini lying around, how can I know if it has this chip or not ? The marking on the chips are a bit erased.

> Your LED strip is very small, the difference is negligible.

The led strip that I will use for HyperHDR is 4m long, 250 leds total. That 1m is additionnal and is just to light up a shelf nearby.

> do not recommend or support either of them.

You recommend HyperSerialWLED then ? I think I'll drop Rpi SPI then, seems too much subject to failures.

Thank you so much for all those hugely useful informations.

**Reply by awawa-dev** [reply]
> Very interesting, thanks a lot for all those infos. I found out that I have an old D1 mini lying around, how can I know if it has this chip or not ? The marking on the chips are a bit erased.

That's not a good sign...probably removed with a laser. Anyway the only way is to test it with 'esp8266/esp32 handshake' enabled to see if HyperHDR detects it over 2Mb speed. Recommend rather HyperSerialEsp8266, WLED from some time has a problem with a serial handling and Wifi handling & LED rendering at the same time. HyperSerialWLED is still based on WLED 0.12, higher version seems not quite stable from my point of view.

> The led strip that I will use for HyperHDR is 4m long, 250 leds total. That 1m is additionnal and is just to light up a shelf nearby.

but it's a different and separate segment? so 250 at once is not a large number.

**Reply by LeVraiRoiDHyrule** [reply]
Hi,
Thanks a lot for your answer, and sorry for the late reply. I've flashed HyperSerialEsp8266 on my wemos.
I've plugged it in and I can see this from the OS:
```
Bus 001 Device 007: ID 1a86:7523 QinHeng Electronics CH340 serial converter
```
I've done the setup as said in the readme and enabled Handshake.
I see this in the logs:
```
2023-05-22T18:59:00.207Z [LEDDEVICE_ADALIGHT] Could not detect HyperSerialEsp8266/HyperSerialESP32 device
2023-05-22T18:59:00.207Z [LEDDEVICE_ADALIGHT] (LedDevice.cpp:333) Switch on
2023-05-22T18:59:00.208Z [LEDDEVICE_ADALIGHT] (LedDevice.cpp:409) Power On
2023-05-22T18:59:00.208Z [LEDDEVICE_ADALIGHT] (LedDevice.cpp:140) Disable device
2023-05-22T18:59:00.208Z [LEDDEVICE_ADALIGHT] (LedDevice.cpp:359) Switch off
2023-05-22T18:59:00.208Z [LEDDEVICE_ADALIGHT] (LedDevice.cpp:317) Set LED strip to black/power off
2023-05-22T18:59:00.208Z [LEDDEVICE_ADALIGHT] (ProviderRs232.cpp:149) Flush was successful
2023-05-22T18:59:00.809Z [LEDDEVICE_ADALIGHT] (ProviderRs232.cpp:131) Close UART: ttyUSB0
2023-05-22T18:59:00.810Z [MUXER0] Priority 240 is now active
2023-05-22T18:59:00.810Z [MUXER0] Set visible priority to 240
2023-05-22T18:59:00.810Z [HYPERHDR0] New priority[240], previous [255]
2023-05-22T18:59:00.811Z [HYPERHDR0] New source available -> switch LED-Device on
2023-05-22T18:59:00.811Z [SMOOTHING0] Clearing queued colors before: enabling
2023-05-22T18:59:00.811Z [SMOOTHING0] Smoothing queue is cleared
2023-05-22T18:59:00.814Z [LEDDEVICE_ADALIGHT] (LedDevice.cpp:103) Enable device
2023-05-22T18:59:00.814Z [LEDDEVICE_ADALIGHT] Opening UART: ttyUSB0
2023-05-22T18:59:00.814Z [LEDDEVICE_ADALIGHT] (ProviderRs232.cpp:216) _rs232Port.open(QIODevice::ReadWrite): ttyUSB0, Baud rate [2000000]bps
2023-05-22T18:59:00.851Z [LEDDEVICE_ADALIGHT] (ProviderRs232.cpp:221) portName:          ttyUSB0
2023-05-22T18:59:00.851Z [LEDDEVICE_ADALIGHT] (ProviderRs232.cpp:222) systemLocation:    /dev/ttyUSB0
2023-05-22T18:59:00.851Z [LEDDEVICE_ADALIGHT] (ProviderRs232.cpp:223) description:       
2023-05-22T18:59:00.851Z [LEDDEVICE_ADALIGHT] (ProviderRs232.cpp:224) manufacturer:      
2023-05-22T18:59:00.851Z [LEDDEVICE_ADALIGHT] (ProviderRs232.cpp:225) productIdentifier: 0x0
2023-05-22T18:59:00.851Z [LEDDEVICE_ADALIGHT] (ProviderRs232.cpp:226) vendorIdentifier:  0x0
2023-05-22T18:59:00.851Z [LEDDEVICE_ADALIGHT] (ProviderRs232.cpp:227) serialNumber:      
2023-05-22T18:59:01.170Z [LEDDEVICE_ADALIGHT] ESP sent: ';7m'$Mp.n'
2023-05-22T18:59:01.478Z [WEBSOCKET] (WebSocketClient.cpp:30) New connection from ::ffff:192.168.1.17
2023-05-22T18:59:01.479Z [WEBSOCKET] (JsonAPI.cpp:97) Client '::ffff:192.168.1.17' switch to HyperHDR instance 0
2023-05-22T18:59:02.188Z [LEDDEVICE_ADALIGHT] Could not detect HyperSerialEsp8266/HyperSerialESP32 device
2023-05-22T18:59:02.188Z [LEDDEVICE_ADALIGHT] (LedDevice.cpp:333) Switch on
2023-05-22T18:59:02.188Z [LEDDEVICE_ADALIGHT] (LedDevice.cpp:409) Power On
```

Does this mean that my Wemos is not compatible ? Could there be another issue ?
Thanks in advance for any answer, have a great day and thanks for your work

EDIT : I've checked again and it seems, according to https://github.com/awawa-dev/HyperHDR/wiki/HyperSerial, that my wemos is not capable of doing HyperSerial.
Would a normal WLED be fast enough then ? I have around 250 leds to control.

**Reply by awawa-dev** [reply]
> 2023-05-22T18:59:01.170Z [LEDDEVICE_ADALIGHT] ESP sent: ';7m'$Mp.n'

yes, it wont work at 2Mb

Cheapest or temporary fastest solution: compile online (read HyperSerial wiki) HyperSerialEsp8266 for 921600 speed and test it again. Use this speed also in HyperHDR instead of 2000000.
Recommended: buy one of supported and well tested esp32 board e.g. ESP32 mh-et live
If you want to test brand new project (HyperSerialPico) buy Raspberry Pi Pico rp2040. These boards are cheap and widely available.
